clients/src/main/java/org/apache/kafka/clients/consumer/internals/StreamsMembershipManager.java

```diff
@@ -0,0 +1,823 @@
+                    activeTasks, standbyTasks, warmupTasks, state);
+                return;
+            }
+
+            processAssignmentReceived(
+                toTasksAssignment(activeTasks),
+                toTasksAssignment(standbyTasks),
+                toTasksAssignment(warmupTasks)
+            );
+        } else {
+            if (responseData.activeTasks() != null ||
+                responseData.standbyTasks() != null ||
+                responseData.warmupTasks() != null) {
+
+                throw new IllegalStateException("Invalid response data, task collections must be all null or all non-null: "
+                    + responseData);
+            }
+        }
+    }
+
+    public void onHeartbeatFailure(boolean retriable) {
+        if (!retriable) {
+            metricsManager.maybeRecordRebalanceFailed();
+        }
+        // The leave group request is sent out once (not retried), so we should complete the leave
+        // operation once the request completes, regardless of the response.
+        if (state == MemberState.UNSUBSCRIBED && maybeCompleteLeaveInProgress()) {
+            log.warn("Member {} with epoch {} received a failed response to the heartbeat to " +
+                "leave the group and completed the leave operation. ", memberIdInfoForLog(), memberEpoch);
+        }
+    }
+
+    public void onPollTimerExpired() {
+        transitionToSendingLeaveGroup(true);
+    }
+
+    public void onFenced() {
+        if (state == MemberState.PREPARE_LEAVING) {
+            log.info("Member {} with epoch {} got fenced but it is already preparing to leave " +
+                "the group, so it will stop sending heartbeat and won't attempt to send the " +
+                "leave request or rejoin.", memberIdInfoForLog(), memberEpoch);
+            finalizeLeaving();
+            transitionTo(MemberState.UNSUBSCRIBED);
+            maybeCompleteLeaveInProgress();
+            return;
+        }
+
+        if (state == MemberState.LEAVING) {
+            log.debug("Member {} with epoch {} got fenced before sending leave group heartbeat. " +
+                "It will not send the leave request and won't attempt to rejoin.", memberIdInfoForLog(), memberEpoch);
+            transitionTo(MemberState.UNSUBSCRIBED);
+            maybeCompleteLeaveInProgress();
+            return;
+        }
+        if (state == MemberState.UNSUBSCRIBED) {
+            log.debug("Member {} with epoch {} got fenced but it already left the group, so it " +
+                "won't attempt to rejoin.", memberIdInfoForLog(), memberEpoch);
+            return;
+        }
+        transitionTo(MemberState.FENCED);
+        resetEpoch();
+        log.debug("Member {} with epoch {} transitioned to {} state. It will release its " +
+            "assignment and rejoin the group.", memberIdInfoForLog(), memberEpoch, MemberState.FENCED);
+
+        CompletableFuture<Void> callbackResult = invokeRevokingAllTasksCallback();
```

Comment:
Is there no difference between revoking all tasksa and losing all tasks?

Reply:
I tried to keep it as simple as possible. If I see a difference when I implement the callbacks on StreamThread-side, I will separate the two events.